src/app/libp2p_helper/src/libp2p_helper/main.go

```diff
@@ -675,40 +676,52 @@ type incomingMsgUpcall struct {
 }
 
 func handleStreamReads(app *app, stream net.Stream, idx int) {
+	// create buffer stream for non-blocking read
+	r := bufio.NewReader(stream)
+
+	var msgBytes []byte
 	go func() {
 		defer func() {
 			_ = stream.Close()
 		}()
-
-		buf := make([]byte, 4096)
 		for {
-			len, err := stream.Read(buf)
-
-			if len != 0 {
-				app.writeMsg(incomingMsgUpcall{
-					Upcall:    "incomingStreamMsg",
-					Data:      codaEncode(buf[:len]),
+			length, err := readLEB128ToUint64(r)
+			if err == io.EOF {
+				continue
+			} else if err != nil {
+				app.writeMsg(streamLostUpcall{
+					Upcall:    "streamLost",
 					StreamIdx: idx,
+					Reason:    fmt.Sprintf("read failure: %s", err.Error()),
 				})
+				return
+			}
+
+			if length == 0 {
+				continue
 			}
 
-			if err != nil && err != io.EOF {
+			msgBytes = make([]byte, length)
+			n, err := io.ReadFull(r, msgBytes)
+			if err != nil {
 				app.writeMsg(streamLostUpcall{
 					Upcall:    "streamLost",
 					StreamIdx: idx,
-					Reason:    fmt.Sprintf("read failure: %s", err.Error()),
+					Reason:    fmt.Sprintf("read failure: %s, read %d bytes", err.Error(), n),
 				})
-				break
+				return
 			}
 
-			if err == io.EOF {
-				break
-			}
+			app.writeMsg(incomingMsgUpcall{
+				Upcall:    "incomingStreamMsg",
+				Data:      codaEncode(msgBytes),
+				StreamIdx: idx,
+			})
+			app.writeMsg(streamReadCompleteUpcall{
```

Comment:
This upcall should be called at the end of each complete message transmission to OCaml. This check should come after an EOF is read.

Reply:
it comes after the `io.ReadFull`, which ensures the entire specified message length is read, so it does send one whole message - also I believe the `if io.EOF {continue}` loop continuously runs until there's more stream data, so not sure if it's good to put it in the `io.EOF` case

Reply:
Would that mean the correct thing to do on `io.EOF` would be to `return`?

Reply:
not necessarily, is there a case where the peers will send multiple messages over the same stream? in that case, it shouldn't return at EOF, because there might be data coming in the future

Reply:
Hmm, this confuses me a little bit. So if we keep the `if err == io.EOF { continue }` behavior here, seems like it would just infinitely loop without pause, since it would go back to the top of the for loop and keep reading `io.EOF` from the reader. Having a hard time finding the semantics of performing a `ReadFull` here wrt EOF, so I might just be missing something.